### PR TITLE
Updating async integration

### PR DESCRIPTION
**Pull request by jasonsandlin**

### CODE REVIEW

File: Tests/LongHaulTestApp/AsyncIntegration.cpp [hunk 1]
```diff
@@ -27,14 +26,121 @@ class win32_handle
...
+            break;
+
+        timer->_End();
+        double timeTaken = timer->max_time();
+        stopAfterMilliseconds -= timeTaken;
+
+        if (stopAfterMilliseconds < 0)
+            break;
+
+        found = true;
```
Comment (brianpepin) [inline Tests/LongHaulTestApp/AsyncIntegration.cpp]: Unnecessary. Would have broken out if found was false.

File: Tests/LongHaulTestApp/AsyncIntegration.cpp [hunk 2]
```diff
@@ -48,66 +154,24 @@ void CALLBACK HandleAsyncQueueCallback(
...
     while (!stop)
     {
         DWORD dwResult = WaitForMultipleObjectsEx(2, hEvents, false, INFINITE, false);
         switch (dwResult)
         {
-        case WAIT_OBJECT_0: // work ready
+        case WAIT_OBJECT_0: 
+            // Background work is ready to be dispatched
             DispatchAsyncQueue(queue, AsyncQueueCallbackType_Work, 0);
             break;
```
Comment (brianpepin) [inline Tests/LongHaulTestApp/AsyncIntegration.cpp]: Need an impl here for WAIT_OBJECT_0 + 1 to set stop.

File: Tests/LongHaulTestApp/AsyncIntegration.cpp [hunk 3]
```diff
@@ -27,14 +26,121 @@ class win32_handle
...
+/// </summary>
 void CALLBACK HandleAsyncQueueCallback(
     _In_ void* context,
     _In_ async_queue_handle_t queue,
     _In_ AsyncQueueCallbackType type
-)
+    )
 {
     UNREFERENCED_PARAMETER(context);
     UNREFERENCED_PARAMETER(queue);
```
Comment (brianpepin) [inline Tests/LongHaulTestApp/AsyncIntegration.cpp]: if (type == AsyncQueueCallbackType_Work) ReleasSemaphore(g_workReadyHandle); ?

File: Tests/LongHaulTestApp/AsyncIntegration.cpp (hunk 1 above)
Comment (jasonsandlin) [reply Tests/LongHaulTestApp/AsyncIntegration.cpp]: oops, accidental test code.  shouldn't be there

---
In reply to: [193533649](https://github.com/Microsoft/xbox-live-api/pull/458#discussion_r193533649) [](ancestors = 193533649)

File: Tests/LongHaulTestApp/AsyncIntegration.cpp (hunk 2 above)
Comment (jasonsandlin) [reply Tests/LongHaulTestApp/AsyncIntegration.cpp]: default will catch it, but i'll add this so its more clear

---
In reply to: [193534220](https://github.com/Microsoft/xbox-live-api/pull/458#discussion_r193534220) [](ancestors = 193534220)

File: Tests/LongHaulTestApp/AsyncIntegration.cpp [hunk 4]
```diff
@@ -27,14 +26,121 @@ class win32_handle
...
+    AddAsyncQueueCallbackSubmitted(queue, nullptr, HandleAsyncQueueCallback, callbackToken);
+
+    ReferenceAsyncQueue(queue); // the BackgroundWorkThreadProc will call close
+    g_hBackgroundWorkThread = CreateThread(nullptr, 0, BackgroundWorkThreadProc, queue, 0, nullptr);
+}
+
+void CleanupAsync(_In_ async_queue_handle_t queue, _In_ uint32_t callbackToken)
+{
+    RemoveAsyncQueueCallbackSubmitted(queue, callbackToken);
+    CloseAsyncQueue(queue);
```
Comment (jplafonta) [inline Tests/LongHaulTestApp/AsyncIntegration.cpp]: set g_stopRequested here?

File: Tests/LongHaulTestApp/AsyncIntegration.cpp (hunk 3 above)
Comment (jasonsandlin) [reply Tests/LongHaulTestApp/AsyncIntegration.cpp]: eh.. near the same perf i think.  and this easier to understand that's its an enum and add the other case if they want to handle it in a single thread.

---
In reply to: [193534501](https://github.com/Microsoft/xbox-live-api/pull/458#discussion_r193534501) [](ancestors = 193534501)

**Pull request merged**